File: elixir/hamming.ex
defmodule Hamming do
  @doc """
  Returns number of differences between two strands of DNA, known as the Hamming Distance.

  ## Examples


  iex> Hamming.hamming_distance('AAGTCATA', 'TAGCGATC')
  {:ok, 4}
  """
  @spec hamming_distance([char], [char]) :: {:ok, non_neg_integer} | {:error, String.t()}
  def hamming_distance(strand1, strand2) when length(strand1) != length(strand2) do
    {:error, "strands must be of equal length"}
  end

  def hamming_distance(strand1, strand2) do
    {:ok,
     Enum.zip_reduce([strand1, strand2], 0, fn
       [c1, c2], acc when c1 != c2 -> acc + 1
       [_c1, _c2], acc -> acc
     end)}
  end
end

File: javascript/hamming.js
export const compute = (strand1, strand2) => {
  if (strand1.length > 0 && strand2.length === 0) {
    throw new Error('right strand must not be empty');
  }

  if (strand1.length === 0 && strand2.length > 0) {
    throw new Error('left strand must not be empty');
  }

  if (strand1.length !== strand2.length) {
    throw new Error('left and right strands must be of equal length');
  }

  return [...strand1].reduce((acc, x, i) => strand2[i] !== x ? acc + 1: acc, 0);
};
